controllers/machineset_controller.go

```diff
@@ -0,0 +1,317 @@
+
+	// Make sure that label selector can match template's labels.
+	selector, err := metav1.LabelSelectorAsSelector(&machineSet.Spec.Selector)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse MachineSet %q label selector: %w", machineSet.Name, err)
+	}
+
+	if !selector.Matches(labels.Set(machineSet.Spec.Template.Labels)) {
+		return nil, fmt.Errorf("failed validation on MachineSet %q label selector, cannot match any machines ", machineSet.Name)
+	}
+
+	// Filter out irrelevant machines (deleting/mismatch labels)
+	machines := []*machinev1.Machine{}
+	for idx := range allMachines.Items {
+		machine := &allMachines.Items[idx]
+		if shouldExcludeMachine(machineSet, machine) {
+			continue
+		}
+
+		machines = append(machines, machine)
+	}
+
+	return machines, nil
+}
+
+func (r ReconcileMachineSet) updateLabelsInMachine(ctx context.Context, machineSet *machinev1.MachineSet, m *machinev1.Machine) error {
+	// Loop through machines and compare labels with machiset
+	if !reflect.DeepEqual(machineSet.Spec.Template.Spec.Labels, m.Spec.Labels) {
+		// Add labels to machine
+		m.Spec.Labels = machineSet.Spec.Template.Spec.Labels
+		err := r.Client.Update(ctx, m)
+		if err != nil {
+			klog.Errorf("failed to update label in %s", m.Name)
+			return err
+		}
+	}
+	return nil
+}
+
+func (r ReconcileMachineSet) updateLabelsInNode(ctx context.Context, m *machinev1.Machine) error {
+	// Get node for machine
+	if m.Status.NodeRef == nil || m.Status.NodeRef.Name == "" {
+		return nil
+	}
+	node := &v1.Node{}
+	err := r.Client.Get(ctx, types.NamespacedName{Name: m.Status.NodeRef.Name}, node)
+	if err != nil {
+		klog.Errorf("failed to fetch node for machine %s", m.Name)
+		return err
+	}
+
+	// Build temp map to store current custom labels in node
+	currentNodeLabels := map[string]string{}
+	// Check node Annotations and compare with Labels to get custom labels
+	currentAnnotationValue, ok := node.Annotations["managed.openshift.com/customlabels"]
+	if ok {
+		for _, lk := range strings.Split(currentAnnotationValue, ",") {
+			if lv, nodeHasLabel := node.Labels[lk]; nodeHasLabel {
+				currentNodeLabels[lk] = lv
+				// Delete label if it's present in node but not in machine
+				if _, machineHasLabel := m.Spec.Labels[lk]; !machineHasLabel {
+					delete(node.Labels, lk)
+				}
+			}
+		}
+	}
+
+	// Compare custom labels with labels in machine
+	if !reflect.DeepEqual(m.Spec.Labels, currentNodeLabels) {
+		// Update Annotations and Labels when new labels are added in machine
+		newAnnotationValue := ""
+		for newKey, newVal := range m.Spec.Labels {
+			node.Labels[newKey] = newVal
+			if newAnnotationValue != "" {
+				newAnnotationValue += ","
+			}
+			newAnnotationValue += newKey
+		}
+		node.Annotations["managed.openshift.com/customlabels"] = newAnnotationValue
+	}
+
+	err = r.Client.Update(ctx, node)
+	if err != nil {
+		klog.Errorf("failed to update label in %s", node.Name)
+		return err
+	}
```

Comment:
wouldn't we always send an update call like that? In the `updateLabelsInMachine` we only run it when there _is_ something to change, should that be the case here as well?